File: calendar_utils.py
"""Calendar utility functions for date calculations and formatting."""

from datetime import date, timedelta
from calendar import monthrange


def format_hour(h: int) -> str:
    if h == 0:
        return "12 AM"
    if h < 12:
        return f"{h} AM"
    if h == 12:
        return "12 PM"
    return f"{h - 12} PM"


def get_hours() -> list[dict[str, int | str]]:
    return [{"hour": h, "label": format_hour(h)} for h in range(6, 24)]


def prev_month(year: int, month: int) -> tuple[int, int]:
    if month == 1:
        return (year - 1, 12)
    return (year, month - 1)


def next_month(year: int, month: int) -> tuple[int, int]:
    if month == 12:
        return (year + 1, 1)
    return (year, month + 1)


def get_month_grid(year: int, month: int) -> list[list[dict]]:
    """Generate a Sunday-start month grid with day info dictionaries."""
    first_day = date(year, month, 1)
    start_dow = first_day.weekday()  # Mon=0 ... Sun=6
    # Convert to Sunday-start: Sun=0, Mon=1, ..., Sat=6
    sun_start_dow = (start_dow + 1) % 7
    start_date = first_day - timedelta(days=sun_start_dow)

    _, days_in_month = monthrange(year, month)
    last_day = date(year, month, days_in_month)

    weeks: list[list[dict]] = []
    current = start_date

    while True:
        week: list[dict] = []
        for _ in range(7):
            week.append({
                "year": current.year,
                "month": current.month,
                "day": current.day,
                "is_current_month": current.month == month and current.year == year,
            })
            current += timedelta(days=1)
        weeks.append(week)

        # Stop after we've passed the last day and current is a Sunday
        # (meaning the week we just built ended on Saturday).
        # Python weekday: Mon=0..Sun=6, so Sunday = 6.
        if current > last_day and current.weekday() == 6:
            break

    return weeks


def get_iso_week(d: date) -> tuple[int, int]:
    """Returns (iso_year, iso_week)."""
    iso = d.isocalendar()
    return (iso[0], iso[1])


def get_week_dates(iso_year: int, iso_week: int) -> list[date]:
    """Get 7 dates for a week, starting from Sunday."""
    # Find the Monday of the given ISO week
    jan4 = date(iso_year, 1, 4)
    # ISO weekday: Mon=1 ... Sun=7
    jan4_iso_dow = jan4.isoweekday()
    monday = jan4 + timedelta(days=(iso_week - 1) * 7 - (jan4_iso_dow - 1))

    # Go back to Sunday for Sunday-start week
    sunday = monday - timedelta(days=1)

    return [sunday + timedelta(days=i) for i in range(7)]


def iso_week_to_date(iso_year: int, iso_week: int) -> date:
    """Returns the Wednesday (mid-week) of the given ISO week."""
    jan4 = date(iso_year, 1, 4)
    jan4_iso_dow = jan4.isoweekday()
    monday = jan4 + timedelta(days=(iso_week - 1) * 7 - (jan4_iso_dow - 1))
    return monday + timedelta(days=2)  # Wednesday


def is_same_day(a: date, b: date) -> bool:
    return a == b


MONTH_NAMES = [
    "", "January", "February", "March", "April", "May", "June",
    "July", "August", "September", "October", "November", "December",
]

DAY_NAMES = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
DAY_ABBREVS_SUN_START = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"]
MONTH_ABBREVS = ["", "Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]


def month_name(month: int) -> str:
    return MONTH_NAMES[month]


def day_name(d: date) -> str:
    # Python weekday: Mon=0..Sun=6, map to names
    names = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
    return names[d.weekday()]


def day_abbrev(d: date) -> str:
    abbrevs = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]
    return abbrevs[d.weekday()]


def format_date(d: date) -> str:
    return f"{day_name(d)}, {MONTH_NAMES[d.month]} {d.day}, {d.year}"


def format_week_label(dates: list[date]) -> str:
    start = dates[0]
    end = dates[6]
    start_str = f"{MONTH_ABBREVS[start.month]} {start.day}"
    if start.month == end.month:
        return f"{start_str} - {end.day}, {end.year}"
    end_str = f"{MONTH_ABBREVS[end.month]} {end.day}, {end.year}"
    return f"{start_str} - {end_str}"

File: calendarUtils.js
const MONTH_NAMES = [
  '', 'January', 'February', 'March', 'April', 'May', 'June',
  'July', 'August', 'September', 'October', 'November', 'December'
];

const MONTH_ABBREVS = [
  '', 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
  'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'
];

const DAY_NAMES = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
const DAY_ABBREVS = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];

function formatHour(h) {
  if (h === 0) return '12 AM';
  if (h < 12) return h + ' AM';
  if (h === 12) return '12 PM';
  return (h - 12) + ' PM';
}

function getHours() {
  const hours = [];
  for (let h = 6; h < 24; h++) {
    hours.push({ hour: h, label: formatHour(h) });
  }
  return hours;
}

function prevMonth(year, month) {
  return month === 1 ? [year - 1, 12] : [year, month - 1];
}

function nextMonth(year, month) {
  return month === 12 ? [year + 1, 1] : [year, month + 1];
}

function daysInMonth(year, month) {
  return new Date(year, month, 0).getDate();
}

function getMonthGrid(year, month) {
  const firstDay = new Date(year, month - 1, 1);
  const startDow = firstDay.getDay(); // Sun=0, Mon=1, ..., Sat=6
  const totalDays = daysInMonth(year, month);

  // Start date: go back startDow days from first of month
  const startDate = new Date(year, month - 1, 1 - startDow);

  const lastDay = new Date(year, month - 1, totalDays);
  const weeks = [];
  const current = new Date(startDate);

  while (true) {
    const week = [];
    for (let i = 0; i < 7; i++) {
      week.push({
        year: current.getFullYear(),
        month: current.getMonth() + 1,
        day: current.getDate(),
        isCurrentMonth: current.getMonth() + 1 === month && current.getFullYear() === year
      });
      current.setDate(current.getDate() + 1);
    }
    weeks.push(week);

    if (current > lastDay && current.getDay() === 0) break;
  }

  return weeks;
}

function getIsoWeek(date) {
  const d = new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate()));
  const dayNum = d.getUTCDay() || 7;
  d.setUTCDate(d.getUTCDate() + 4 - dayNum);
  const yearStart = new Date(Date.UTC(d.getUTCFullYear(), 0, 1));
  const weekNo = Math.ceil(((d - yearStart) / 86400000 + 1) / 7);
  return [d.getUTCFullYear(), weekNo];
}

function getWeekDates(isoYear, isoWeek) {
  const jan4 = new Date(isoYear, 0, 4);
  const jan4Dow = jan4.getDay() || 7; // Mon=1..Sun=7
  const mondayOffset = (isoWeek - 1) * 7 - (jan4Dow - 1);
  const monday = new Date(isoYear, 0, 4 + mondayOffset);
  const sunday = new Date(monday);
  sunday.setDate(monday.getDate() - 1);

  const dates = [];
  for (let i = 0; i < 7; i++) {
    const d = new Date(sunday);
    d.setDate(sunday.getDate() + i);
    dates.push(d);
  }
  return dates;
}

function isoWeekToDate(isoYear, isoWeek) {
  const jan4 = new Date(isoYear, 0, 4);
  const jan4Dow = jan4.getDay() || 7;
  const mondayOffset = (isoWeek - 1) * 7 - (jan4Dow - 1);
  const monday = new Date(isoYear, 0, 4 + mondayOffset);
  const wednesday = new Date(monday);
  wednesday.setDate(monday.getDate() + 2);
  return wednesday;
}

function monthName(month) {
  return MONTH_NAMES[month];
}

function dayName(date) {
  return DAY_NAMES[date.getDay()];
}

function dayAbbrev(date) {
  return DAY_ABBREVS[date.getDay()];
}

function formatDate(d) {
  return `${dayName(d)}, ${MONTH_NAMES[d.getMonth() + 1]} ${d.getDate()}, ${d.getFullYear()}`;
}

function formatWeekLabel(dates) {
  const start = dates[0];
  const end = dates[6];
  const startStr = `${MONTH_ABBREVS[start.getMonth() + 1]} ${start.getDate()}`;
  if (start.getMonth() === end.getMonth()) {
    return `${startStr} - ${end.getDate()}, ${end.getFullYear()}`;
  }
  const endStr = `${MONTH_ABBREVS[end.getMonth() + 1]} ${end.getDate()}, ${end.getFullYear()}`;
  return `${startStr} - ${endStr}`;
}

function sameDay(a, b) {
  return a.getFullYear() === b.getFullYear()
    && a.getMonth() === b.getMonth()
    && a.getDate() === b.getDate();
}

function navContext(viewMode, ctxYear, ctxMonth, ctxDay, ctxIsoYear, ctxIsoWeek) {
  const today = new Date();
  const todayIso = getIsoWeek(today);
  return {
    viewMode,
    ctxYear,
    ctxMonth,
    ctxDay,
    ctxIsoYear,
    ctxIsoWeek,
    today,
    todayIsoYear: todayIso[0],
    todayIsoWeek: todayIso[1]
  };
}

module.exports = {
  formatHour,
  getHours,
  prevMonth,
  nextMonth,
  getMonthGrid,
  getIsoWeek,
  getWeekDates,
  isoWeekToDate,
  monthName,
  dayName,
  dayAbbrev,
  formatDate,
  formatWeekLabel,
  sameDay,
  navContext,
  DAY_HEADERS: ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat']
};
